test/mongodb.test.js

```diff
@@ -1032,6 +1032,16 @@ describe('mongodb connector', function() {
   });
 
   describe('updateAll', function() {
+    it('should preserve the id', async function() {
+      let user = await User.create({name: 'Al', age: 31, email: 'al@strongloop'});
+      const userId = user.id;
+      user = user.toObject();
+      user.age = 100;
+
+      await User.update(user);
+      user.should.have.property('id', userId);
```

Comment:
What is the expected `id` type - is it a string or `ObjectID` instance?

Also, `user` is the object given to `User.update`. Usually, methods don't modify objects passed in arguments, therefore I would expect that `user.id` remains the same after `User.update` as it was before making the call.

I.e. this test can be simplified as follows:

```ts
const userId = user.id;
user = user.toObject();
user.should.have.property('id', userId);
```

I think you should check `id` property in the value returned by `User.update` instead?

```ts
const result = await User.update(user);
result.should.have.property('id', userId);
```

The data returned by DAO methods like `update` don't necessarily match the data stored in the database. To be safe, you may want to check database records too.

```ts
const stored = await User.findById(userId);
stored.should.have.property('id', userId);
```

Reply:
I don't know about others, but this is the behavior of the Mongo connector on doing `User.update(user)`.

1. Modifies `user`
2. Returns an object with a `count` property, which is the number of items affected, does not return the model or a `toObject()` representation of it.

https://github.com/strongloop/loopback-next/issues/3267 is the result of modifying `user`.

`user` is not modified if it is an actual instance of a Model. However, `repository.create()` returns a `toObject()` version of the model, which the user passes to `repository.update()`, hence the modification.

Reply:
That makes we wonder why should `repository.create()` not return the model instance?

Reply:
> Returns an object with a count property, which is the number of items affected, does not return the model or a toObject() representation of it.

Ah, good point, I never remember which method returns `{count}` and which returns the updated model instance. Thanks for clarifying!

> Modifies `user` if it is not an instance of a Model.

While `user` parameter can be modified inside a method, this change is not visible to the caller.

```js
const x = {orig: true};
function f(data) { data = {modified: true}; }
f(x);
console.log(x);
// prints: { orig: true }
```

Reply:
@hacksparrow this discussion thread is not resolved.

See also Raymond's comment below:

> The test name is misleading. IIUC, it is testing `update() does not mutate the input model object`.

I think you need to decide what is the scenario you are testing and then rework & clarify the tests to test only that scenario and nothing else.